README

                           Archive-Zip-SimpleZip

                               Version 0.039

                              8 September 2020

        Copyright (c) 2012-2020 Paul Marquess. All rights reserved.
          This program is free software; you can redistribute it
           and/or modify it under the same terms as Perl itself.


DESCRIPTION
-----------

This distribution provides a Perl interface to allow reading and writing of
Zip files.


PREREQUISITES
-------------

Before you can build Archive-Zip-SimpleZip you need to have the following
installed on your system:

    * Perl 5.006 or better.
    * IO::Compress version 2.096 or better.

BUILDING THE MODULE
-------------------

Assuming you have met all the prerequisites, the module can now be built
using this sequence of commands:

    perl Makefile.PL
    make
    make test

INSTALLATION
------------

To install Archive::Zip::SimpleZip, run the command below:

    make install

SUPPORT
-------

General feedback/questions/bug reports should be sent to
https://github.com/pmqs/Archive-Zip-SimpleZip/issues (preferred) or
https://rt.cpan.org/Public/Dist/Display.html?Name=Archive-Zip-SimpleZip.

FEEDBACK
--------

You can report a problem with Archive-Zip-SimpleZip to

    https://github.com/pmqs/Archive-Zip-SimpleZip/issues

The folowing address is still active, but the github address above is preferred.

    https://rt.cpan.org/Public/Dist/Display.html?Name=Archive-Zip-SimpleZip

To help me help you, I need all of the following information:

 1. The Versions of everything relevant.
    This includes:

     a. The *complete* output from running this

            perl -V

        Do not edit the output in any way.
        Note, I want you to run "perl -V" and NOT "perl -v".

        If your perl does not understand the "-V" option it is too
        old. This module needs Perl version 5.006 or better.

     b. The version of Archive-Zip-SimpleZip you have.
        If you have successfully installed Archive-Zip-SimpleZip, this one-liner
        will tell you:

           perl -MArchive::Zip::SimpleZip -e 'print qq[ver $Archive::Zip::SimpleZip::VERSION\n]'

        If you are  running windows use this

           perl -MArchive::Zip::SimpleZip -e "print qq[ver $Archive::Zip::SimpleZip::VERSION\n]"

        If you haven't installed Archive-Zip-SimpleZip then search Archive::Zip::SimpleZip
        for a line like this:

          $VERSION = "2.052" ;

 2. If you are having problems building Archive-Zip-SimpleZip, send me a
    complete log of what happened. Start by unpacking the Archive-Zip-SimpleZip
    module into a fresh directory and keep a log of all the steps

        [edit config.in, if necessary]
        perl Makefile.PL
        make
        make test TEST_VERBOSE=1

Paul Marquess <pmqs@cpan.org>


Changes
CHANGES
-------

  0.037 4 September 2020

        * Support for XZ & Zstd compression
          17f976cfea7169b487d044de5c2ace42058cdd5c

        * trim whitespace
          4951bf18e9bca1b0ee986d016f46c0134c9e2759

  0.036 3 August 2020

        * 111const-deflate.t - fix number of tests
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/13
          303259e977dad9785cacfd798c3d3e19921f90f0

  0.035 7 December 2019

        * 000prereq.t - needs version 2.092 of dependencies
          Addresses smokre failure with t/010main-unzip.t
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/7
          ea4bfdd14d4e0047d8e3fe55c6ce88de2e1ccab0

  0.034 5 December 2019

        * 000prereq.t - needs version 2.092 of dependencies
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/11
          330dfb7dda1b3016eec52dc1c90ff30110aa8f6a

  0.033 4 December 2019

        * fix typo
          88221114c4edf938ad506845db9863db17f7492e

        * Makefile.PL: Add IO::Compress::Lzma to PREREQ_PM
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/9
          896604d5e525a22f25877f5595ffa94ffdd8a6cf

  0.032 24 November 2019

        * Add StreamedUnzip
          6d2c59aa70f5992903fdeb5d8a3b7fc97dc84137

        * fix typo in Changes
          2c4bd7743905ba04f121bf323d09caf8f6e88181

  0.031 10 November 2019

        * 000prereq.t: Add more prereqs for LZMA
          If have the IO::*:: LZma modules, they need to be up to date
          Also print IO::*::Lzma module versions
          318960e89674b2c3a450db972e8f27b14dd1d685

        * Makefile.PL: Add IO::*::Lzma to "recommends"  list
          8675d2421adc7c846a58cf6ddbffe255cf61a21d

  0.030 9 November 2019

        * t/010main-unzip.t fails
          Issue triggered by a memory leak in Compress-Raw-Lzma.
          Updating PREREQ_PM will force loading of fixed Compress-Raw-Lzma
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/7

  0.029 8 November 2019

        * Fix Use of uninitialized value error
          53fee715828dfc08071b34fffb7408de91758d09

  0.028 7 November 2019

        * Beef up test for library versions
          a154c5c7e286286c393e7b1832b77e0ca48d5594

  0.027 7 November 2019

        * t/000prereq.t
          Print our versions of all modules used
          c560193b1254d8515f2d880a713ba1ece888c0b9

  0.026 5 November 2019

        * Can't extract after read content
          Fixes to allow nembers to be read multiple times
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/3
          76943104cfce9e3ee4a4772b115fdb89faefa6fc

        * add example of creating zip from stdin
          bfb5b815be3f7d5a1600a48509272da37a6d8b3d

        * use mkpath in SimpleUnzip.pm
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/2
          d683c1d5f5e3c91bfd6beac3187cda6ca232d22c

        * uncompr error with content()
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/4
          efede5832f33b403c767a458c22fbc7d1f21c672

        * remove bad test
          9e302ae233c17b70699091fdd947f2fd2b4c0cfa

        * update to chmod for read-only
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/2
          28135fa2228790e920e102fd444606c926444246

        * Error when i try extract file to full path
          https://github.com/pmqs/Archive-Zip-SimpleZip/issues/2
          6b33be16c82fab8fe3f85b66c2bce21424960c55

        * Add Support section
          50190d09901a5a188ad308e8a015563b66e65ca6

        * more tests with deflated directory
          c8c15cf3f0c47eaf9d9490aa508f14e184b1ba02

        * test with a deflated directory
          d3cc32b2cc92dba64ba879320b80d6696e56f62a

        * Documentation Updates
          0978433940737981a97dc1cff083ba4a1dbb4dd3

  0.025 19 June 2019

        * Don't use // for older perls
          9f9462be5fe6f6356406ff7dcd1b3e32f6ebebb7

        * Add Perl::OSType as a prerequisite module
          22aeb3f6615752602c48942f1a0e822ed3cb9d1c

  0.024 16 June 2019

        * Respin release

  0.023 16 June 2019

      * SimpleZip.pm
        Populate extended attributes for file times & unix uid/gid
        2036dcc60355ad32d6e1d3dbc5cbe761327e967f

      * SimpleUnzip.pm
        * Added FilesOnly option
        * Added extract method
          0f1e98f94c6e0c8a8a874311bcdc18cc90ea7b54

      * Added source to github

      * Added .travis & .appveyor files

  0.022 10 April 2018

      *  t/010main-unzip.t
         Fixed failure on Windows

  0.021 8 April 2018

      * Makefile.PL
        Fixed PREREQ_PM on IO::Compress::Zip  - should be version 2.081
        [#125043]

      *  t/010main-unzip.t
         test may fail with non-English locale
         [# 125044]


  0.020 2 April 2018

      * added Archive::Zip::SimpleUnzip

  0.010 10 April 2017

      * Makefile.PL
        #120084: Need Fix for Makefile.PL depending on . in @INC

      * added autoflush option.

      * croak if attempt to freeze/thaw SimpleZip object

  0.009 17 February 2013

      * added addFileHandle method.

      * beefed up addMember

  0.008 7 January 2013

      * Fixes to allow disabling of CanonicalName

      * Document how to use File::Find to zip a directory tree.

  0.007 17 December 2012

      * Removed non-ASIII char that triggered failure in t/999pod.t.

  0.006 10 December 2012

      * Added openMember to allow filehandle access to writing zip files.

      * Added Encode option.

  0.005 10 November 2012

      * Performance improvements.

  0.004 11 August 2012

      * Force binmode for input file.

  0.003 1 May 2012

      * Made 'Name' option mandatory with addString.

  0.002 26 April 2012

      * Fixed issue with Makefile.PL & version dependencies.

  0.001 26 April 2012

      * First Release


lib/Archive/Zip/StreamedUnzip.pm
package Archive::Zip::StreamedUnzip;

require 5.006;

use strict ;
use warnings;
use bytes;

use IO::File;
use Carp;
use Scalar::Util ();

use IO::Compress::Base::Common  2.093 qw(:Status);
use IO::Compress::Zip::Constants 2.093 ;
use IO::Uncompress::Unzip 2.093 ;


require Exporter ;

our ($VERSION, @ISA, @EXPORT_OK, %EXPORT_TAGS, $StreamedUnzipError);

$VERSION = '0.001';
$StreamedUnzipError = '';

@ISA    = qw(IO::Uncompress::Unzip Exporter);
@EXPORT_OK = qw( $StreamedUnzipError unzip );
%EXPORT_TAGS = %IO::Uncompress::RawInflate::EXPORT_TAGS ;
push @{ $EXPORT_TAGS{all} }, @EXPORT_OK ;
Exporter::export_ok_tags('all');


sub _setError
{
    $StreamedUnzipError = $_[2] ;
    $_[0]->{Error} = $_[2]
        if defined  $_[0] ;

    return $_[1];
}

sub _illegalFilename
{
    return _setError(undef, undef, "Illegal Filename") ;
}

sub is64BitPerl
{
    use Config;
    # possibly use presence of pack/unpack "Q" for int size test?
    $Config{lseeksize} >= 8 and $Config{uvsize} >= 8;
}

sub new
{
    my $class = shift ;

    return _setError(undef, undef, "Missing Filename")
        unless @_ ;

    my $inValue = shift ;
    my $fh;

    if (!defined $inValue)
    {
        return _illegalFilename
    }

    my $isSTDOUT = ($inValue eq '-') ;
    my $inType = IO::Compress::Base::Common::whatIsOutput($inValue);

    if ($inType eq 'filename')
    {
        if (-e $inValue && ( ! -f _ || ! -r _))
        {
            return _illegalFilename
        }

        $fh = new IO::File "<$inValue"
            or return _setError(undef, undef, "cannot open file '$inValue': $!");
    }
    elsif( $inType eq 'buffer' || $inType eq 'handle')
    {
        $fh = $inValue;
    }
    else
    {
        return _illegalFilename
    }

    my %obj ;
    # my $inner = IO::Compress::Base::Common::createSelfTiedObject($class, \$StreamedUnzipError);

    # # *$inner->{Pause} = 1;
    # $inner->_create(undef, 0, $fh, @_)
    #     or return undef;

    my $inner = IO::Uncompress::Unzip->new($fh) ;

    $obj{Inner} = $inner;
    $obj{Open} = 1 ;
    $obj{FirstOne} = 1 ;

    bless \%obj, $class;
}

sub close
{
    my $self = shift;
    # TODO - fix me
    $self->{Inner}->close();
    return 1;
}

sub DESTROY
{
    my $self = shift;
}

sub next
{
    my $self = shift;

    if ($self->{FirstOne})
    {
        $self->{FirstOne} = 0;
    }
    else
    {
        my $status = $self->{Inner}->nextStream();
        return undef
            if $status <= 0;
    }

    my %member ;
    $member{Inner}  = $self->{Inner};
    # $member{Member} = $member;
    $member{Info} = $self->{Inner}->getHeaderInfo() ;
    #Scalar::Util::weaken $member{Inner}; # for 5.8

    return bless \%member, 'Archive::Zip::StreamedUnzip::Member';
}

sub member
{
    my $self = shift;
    my $name = shift;

    return _setError(undef, undef, "Member '$name' not in zip")
        if ! defined $name ;

    while (my $member = $self->next())
    {
        return $member
            if $member->name() eq $name ;

    }

    return _setError(undef, undef, "Member '$name' not in zip") ;
}

sub getExtraParams
{

    return (
            # Zip header fields
            'name'    => [IO::Compress::Base::Common::Parse_any,       undef],

#            'stream'  => [IO::Compress::Base::Common::Parse_boolean,   1],
        );
}

sub ckParams
{
    my $self = shift ;
    my $got = shift ;

    # unzip always needs crc32
    $got->setValue('crc32' => 1);

    *$self->{UnzipData}{Name} = $got->getValue('name');

    return 1;
}




{
    package Archive::Zip::StreamedUnzip::Member;

    sub name
    {
        my $self = shift;
#        $self->_stdPreq() or return 0 ;

        return $self->{Info}{Name};
    }

    sub isDirectory
    {
        my $self = shift;
#        $self->_stdPreq() or return 0 ;

        return substr($self->{Info}{Name}, -1, 1) eq '/' ;
    }

    sub isFile
    {
        my $self = shift;
#        $self->_stdPreq() or return 0 ;

        # TODO - test for symlink
        return ! $self->isDirectory() ;
    }

# TODO
#
#    isZip64
#    isDir
#    isSymLink
#    isText
#    isBinary
#    isEncrypted
#    isStreamed
#    getComment
#    getExtra
#    compressedSize - 64 bit alert
#    uncompressedSize
#    time
#    isStored
#    compressionName
#
#   extractToFile

    sub compressedSize
    {
        my $self = shift;
#        $self->_stdPreq() or return 0 ;

        my $CompressedLength = $self->{Info}{CompressedLength};
        if (ref $CompressedLength)
        {
            return U64::get64bit($CompressedLength)
        }
        return $CompressedLength;
    }

    sub uncompressedSize
    {
        my $self = shift;
#        $self->_stdPreq() or return 0 ;
        my $UncompressedLength = $self->{Info}{UncompressedLength};
        if (ref $UncompressedLength)
        {
            return U64::get64bit($UncompressedLength)
        }
        return $UncompressedLength;
    }

    sub content
    {
        my $self = shift;
        my $data ;

        # $self->{Inner}->read($data, $self->{UncompressedLength});
        $self->{Inner}->read($data, $self->{Info}{UncompressedLength});

        return $data;
    }

    sub open
    {
        my $self = shift;

#        return  return $self->{Inner} ;

#        my $handle = Symbol::gensym();
#        tie *$handle, "Archive::Zip::StreamedUnzip::Handle", $self->{SZ}{UnZip};
#        return $handle;

        my $z = IO::Compress::Base::Common::createSelfTiedObject("Archive::Zip::StreamedUnzip::Handle", \$StreamedUnzipError) ;

        *$z->{Open} = 1 ;
        *$z->{SZ} = $self->{Inner};
        Scalar::Util::weaken *$z->{SZ}; # for 5.8

        $z;
    }

    sub close
    {
        my $self = shift;
        return 1;
    }


}


{
    package Archive::Zip::StreamedUnzip::Handle ;

    sub TIEHANDLE
    {
        return $_[0] if ref($_[0]);
        die "OOPS\n" ;
    }

    sub UNTIE
    {
        my $self = shift ;
    }

    sub DESTROY
    {
#        print "DESTROY H";
        my $self = shift ;
        local ($., $@, $!, $^E, $?);
        $self->close() ;

        # TODO - memory leak with 5.8.0 - this isn't called until
        #        global destruction
        #
        %{ *$self } = () ;
        undef $self ;
    }


    sub close
    {
        my $self = shift ;
        return 1 if ! *$self->{Open};

        *$self->{Open} = 0 ;

#        untie *$self
#            if $] >= 5.008 ;

        if (defined *$self->{SZ})
        {
#            *$self->{SZ}{Raw} = undef ;
            *$self->{SZ} = undef ;
        }

        1;
    }

    sub read
    {
        # TODO - remember to fix the return value to match real read & not the broken one in IO::Uncompress
        my $self = shift;
        $self->_stdPreq() or return 0 ;

#        warn "READ [$self]\n";
#        warn "READ [*$self->{SZ}]\n";

#        $_[0] = *$self->{SZ}{Unzip};
#        my $status = goto &IO::Uncompress::Base::read;
#        $_[0] = \$_[0] unless ref $_[0];
        my $status = *$self->{SZ}->read(@_);
        $status = undef if $status < 0 ;
        return $status;
    }

    sub readline
    {
        my $self = shift;
        $self->_stdPreq() or return 0 ;
        *$self->{SZ}->getline(@_);
    }

    sub tell
    {
        my $self = shift;
        $self->_stdPreq() or return 0 ;

        *$self->{SZ}->tell(@_);
    }

    sub eof
    {
        my $self = shift;
        $self->_stdPreq() or return 0 ;

        *$self->{SZ}->eof;
    }

    sub _stdPreq
    {
        my $self = shift;

        # TODO - fix me
        return 1;

        return _setError("Zip file closed")
            if ! defined defined *$self->{SZ} || ! *$self->{Inner}{Open} ;


        return _setError("member filehandle closed")
            if  ! *$self->{Open} ; #|| ! defined *$self->{SZ}{Raw};

        return 0
            if *$self->{SZ}{Error} ;

         return 1;
    }

    sub _setError
    {
        $Archive::Zip::SimpleUnzip::StreamedUnzipError = $_[0] ;
        return 0;
    }

    sub clearerr
    {
        my $self = shift;

        return 0;
    }

    sub binmode { 1 }

#    sub clearerr { $Archive::Zip::SimpleUnzip::StreamedUnzipError = '' }

    *BINMODE  = \&binmode;
#    *SEEK     = \&seek;
    *READ     = \&read;
    *sysread  = \&read;
    *TELL     = \&tell;
    *READLINE = \&readline;
    *EOF      = \&eof;
    *FILENO   = \&fileno;
    *CLOSE    = \&close;
}


1;

__END__

=head1 NAME

Archive::Zip::StreamedUnzip - Read Zip Archives in streaming mode

=head1 SYNOPSIS

    use Archive::Zip::StreamedUnzip qw($StreamedUnzipError) ;

    my $z = new Archive::Zip::StreamedUnzip "my.zip"
        or die "Cannot open zip file: $StreamedUnzipError\n" ;


    # Iterate through a zip archive
    while (my $member = $z->next)
    {
        print $member->name() . "\n" ;
    }

    # Archive::Zip::StreamedUnzip::Member

    my $name = $member->name();
    my $content = $member->content();
    my $comment = $member->comment();

    # open a filehandle to read from a zip member
    $fh = $member->open("mydata1.txt");

    # read blocks of data
    read($fh, $buffer, 1234) ;

    # or a line at a time
    $line = <$fh> ;

    close $fh;

    $z->close();

=head1 DESCRIPTION

Archive::Zip::StreamedUnzip is a module that allows reading of Zip archives in streaming mode.
This is useful if you are processing a zip coming directly off a socket without having to
read the complete file into memory and/or store it on disk. Similarly it can be handy when
woking with a pipelined command.

Working with a streamed zip file does have limitations, so
most of the time L<Archive::Zip::SimpleUnzip> and/or L<Archive::Zip> are a  better choice of
module for reading file files.

For writing Zip archives, there is a companion module,
called L<Archive::Zip::SimpleZip>, that can create Zip archives.

B<NOTE> This is alpha quality code, so the interface may change.

=head2 Features

=over 5

=item * Read zip archive from a file, a filehandle or from an in-memory buffer.

=item * Perl Filehandle interface for reading a zip member.

=item * Supports deflate, store, bzip2 and lzma compression.

=item * Supports Zip64, so can read archves larger than 4Gig and/or have greater than 64K members.

=back

=head2 Constructor

     $z = new Archive::Zip::StreamedUnzip "myzipfile.zip" [, OPTIONS] ;
     $z = new Archive::Zip::StreamedUnzip \$buffer [, OPTIONS] ;
     $z = new Archive::Zip::StreamedUnzip $filehandle [, OPTIONS] ;

The constructor takes one mandatory parameter along with zero or more
optional parameters.

The mandatory parameter controls where the zip archive is read from.
This can be any one of the following

=over 5

=item * Input from a Filename

When StreamedUnzip is passed a string, it will read the zip archive from the
filename stored in the string.

=item * Input from a String

When StreamedUnzip is passed a string reference, like C<\$buffer>, it will
read the in-memory zip archive from that string.

=item * Input from a Filehandle

When StreamedUnzip is passed a filehandle, it will read the zip archive from
that filehandle. Note the filehandle must be seekable.


=back

See L</Options> for a list of the optional parameters that can be specified
when calling the constructor.

=head2 Options

None yet.

=head2 Methods

=over 5

=item $z->next()

Returns the next member from the zip archive as a
Archive::Zip::StreamedUnzip::Member object.
See L</Archive::Zip::StreamedUnzip::Member>

Standard usage is

    use Archive::Zip::StreamedUnzip qw($StreamedUnzipError) ;

    my $match = "hello";
    my $zipfile = "my.zip";

    my $z = new Archive::Zip::StreamedUnzip $zipfile
        or die "Cannot open zip file: $StreamedUnzipError\n" ;

    while (my $member = $z->next())
    {
        my $name = $member->name();
        my $fh = $member->open();
        while (<$fh>)
        {
            my $offset =
            print "$name, line $.\n" if /$match/;
        }
    }

=item $z->close()

Closes the zip file.

=back

=head1 Archive::Zip::StreamedUnzip::Member

The C<next> method returns a member object of
type C<Archive::Zip::StreamedUnzip::Member>
that has the following methods.

=over 5

=item $string = $m->name()

Returns the name of the member.

=item $data = $m->content()

Returns the uncompressed content.

=item $fh = $m->open()

Returns a filehandle that can be used to read the uncompressed content.

=back

=head1 Examples

=head2 Iterate through a Zip file

    use Archive::Zip::StreamedUnzip qw($StreamedUnzipError) ;

    my $zipfile = "my.zip";
    my $z = new Archive::Zip::StreamedUnzip $zipfile
        or die "Cannot open zip file: $StreamedUnzipError\n" ;

    while (my $member = $z->next())
    {
        print "$member->name()\n";
    }

=head2 Filehandle interface

Here is a simple grep, that walks through a zip file and
prints matching strings.

    use Archive::Zip::StreamedUnzip qw($StreamedUnzipError) ;

    my $match = "hello";
    my $zipfile = "my.zip";

    my $z = new Archive::Zip::StreamedUnzip $zipfile
        or die "Cannot open zip file: $StreamedUnzipError\n" ;

    while (my $member = $z->next())
    {
        my $name = $member->name();
        my $fh = $member->open();
        while (<$fh>)
        {
            my $offset =
            print "$name, line $.\n" if /$match/;
        }
    }


=head2 Nested Zip

Here is a script that will list the contents of a zip file along with any zip files that are embedded in it.
In fact it will work with any level of nesting.

    sub walk
    {
        my $unzip  = shift ;
        my $depth = shift // 1;

        while (my $member = $unzip->next())
        {
            my $name = $unzip->name();
            print "  " x $depth . "$name\n" ;

            if ($name =~ /\.zip$/i)
            {
                my $fh = $member->open();
                my $newunzip = new Archive::Zip::StreamedUnzip $fh;
                walk($newunzip, $depth + 1);
            }
        }
    }

    my $unzip = new Archive::Zip::StreamedUnzip $zipfile
                or die "Cannot open '$zipfile': $StreamedUnzipError";

    print "$zipfile\n" ;
    walk($unzip) ;

=head1 Zip File Interoperability

The intention is to be interoperable with zip archives created by other
programs, like pkzip or WinZip, but the majority of testing carried out
used the L<Info-Zip zip/unzip|http://www.info-zip.org/> programs
running on Linux.

This doesn't necessarily mean that there is no interoperability with other
zip programs like pkzip and WinZip - it just means that I haven't tested
them. Please report any issues you find.

=head2 Compression Methods Supported

The following compression methods are supported

=over 5

=item deflate (8)

This is the most common compression used in zip archives.

=item store (0)

This is used when no compression has been carried out.

=item bzip2 (12)

Only if the C<IO-Compress-Bzip2> module is available.

=item lzma (14)

Only if the C<IO-Compress-Lzma> module is available.

=back

=head2 Zip64 Support

This modules supports Zip64, so it can read archves larger than 4Gig
and/or have greater than 64K members.


=head2 Limitations

The following features are not currently supported.

=over 4

=item * Compression methods not listed in L</Compression Methods Supported>

=item * Multi-Volume Archives

=item * Encrypted Archives

=back

=head1 SUPPORT

General feedback/questions/bug reports should be sent to
L<https://github.com/pmqs/Archive-Zip-SimpleZip/issues> (preferred) or
L<https://rt.cpan.org/Public/Dist/Display.html?Name=Archive-Zip-SimpleZip>.


=head1 SEE ALSO


L<Archive::Zip::SimpleUnzip>, L<Archive::Zip::SimpleZip>, L<Archive::Zip>, L<IO::Compress::Zip>,  L<IO::Uncompress::UnZip>


=head1 AUTHOR

This module was written by Paul Marquess, F<pmqs@cpan.org>.

=head1 MODIFICATION HISTORY

See the Changes file.

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2019 Paul Marquess. All rights reserved.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

Makefile.PL
#! perl -w

use strict ;
require 5.006 ;

use lib '.';
use private::MakeUtil;
use ExtUtils::MakeMaker 5.16 ;

my $ParentVersion = 2.096 ;

# Check if th eother IO::Compress::* modules are installed.
# If any are, add to PREREQ_PM.
my %extra_prereq = ();
$extra_prereq{'IO::Compress::Lzma'} = $ParentVersion
    if eval { require IO::Compress::Lzma; 1 } ;
$extra_prereq{'IO::Compress::Bzip2'} = $ParentVersion
    if eval { require IO::Compress::Bzip2; 1 } ;
$extra_prereq{'IO::Compress::Xz'} = $ParentVersion
    if eval { require IO::Compress::Xz; 1 } ;
$extra_prereq{'IO::Compress::Zstd'} = $ParentVersion
    if eval { require IO::Compress::Zstd; 1 } ;

WriteMakefile(
    NAME         => 'Archive::Zip::SimpleZip',
    VERSION_FROM => 'lib/Archive/Zip/SimpleZip.pm',
    'dist'       => { COMPRESS     => 'gzip',
                      TARFLAGS     => '-chvf',
                      SUFFIX       => 'gz',
                      DIST_DEFAULT => 'MyTrebleCheck tardist',
                    },


    PREREQ_PM   => {
            'IO::Compress::Base'   => $ParentVersion,
            'IO::Uncompress::Base' => $ParentVersion,
            'Perl::OSType'         => 0,
            %extra_prereq,
            } ,

    ( $] >= 5.005
        ? (ABSTRACT => 'Read and Write Zip Archives',
           AUTHOR   => 'Paul Marquess <pmqs@cpan.org>')
        : ()
    ),

    ( eval { ExtUtils::MakeMaker->VERSION(6.31) }
        ? ('LICENSE'  => 'perl')
        : ()
    ),

     ( eval { ExtUtils::MakeMaker->VERSION(6.46) }
        ? ( META_MERGE  => {

                "meta-spec" => { version => 2 },

                no_index => {
                    directory => [ 't', 'private' ],
                },

                resources   => {

                    bugtracker  => {
                        web     => 'https://github.com/pmqs/Archive-Zip-SimpleZip/issues'
                    },

                    homepage    => 'https://github.com/pmqs/Archive-Zip-SimpleZip',

                    repository  => {
                        type    => 'git',
                        url     => 'git://github.com/pmqs/Archive-Zip-SimpleZip.git',
                        web     => 'https://github.com/pmqs/Archive-Zip-SimpleZip',
                    },
                },

                prereqs => {
                    runtime => {
                        requires => {
                            'IO::Compress::Base'        => $ParentVersion,
                            'IO::Uncompress::Base'      => $ParentVersion,
                            'IO::Compress::Zip'         => $ParentVersion,
                            'IO::Uncompress::Umzip'     => $ParentVersion,
                            'IO::Compress::RawDeflate'  => $ParentVersion,
                            'IO::Uncompress::RawInflate'=> $ParentVersion,
                            'Perl::OSType'              => 0,
                        },
                        recommends => {
                            'IO::Compress::Lzma'        => $ParentVersion,
                            'IO::Uncompress::UnLzma'    => $ParentVersion,
                            'IO::Compress::Zstd'        => $ParentVersion,
                            'IO::Uncompress::UnZstd'    => $ParentVersion,
                            'IO::Compress::Bzip2'       => $ParentVersion,
                            'IO::Uncompress::Bunzip2'   => $ParentVersion,
                            'IO::Compress::Xz'          => $ParentVersion,
                            'IO::Uncompress::UnXz'      => $ParentVersion,
                        },
                    },
                },
              }
            )
        : ()
    ),
) ;

# end of file Makefile.PL
